criu/pages-compress.c

```diff
@@ -0,0 +1,285 @@
+	const char *compbuf = mapped_addr;
+	const int outbufsize = 4 * 1024 * 1024; // LZ4F_max4MB is the max block size; see LZ4F_blockSizeID_t for details
+	char outbuf[outbufsize];
+	LZ4F_errorCode_t lz4err;
+	int decomp_fd;
+	size_t totalread = 0;
+	size_t totalwrite = 0;
+
+	snprintf(decompressed_filename, sizeof(decompressed_filename), decompressed_filename_format, (int)getpid());
+	pr_debug("Creating tmp file at '%s'\n", decompressed_filename);
+	decomp_fd = open(decompressed_filename, O_CREAT | O_WRONLY | O_TRUNC, S_IRUSR | S_IWUSR);
+
+	if (0 > decomp_fd) {
+		pr_err("Can't create file, errno=%d\n", errno);
+		return -1;
+	}
+
+	while (1) {
+		size_t insize = mapped_size - (compbuf - mapped_addr);
+		if (!insize) {
+			/* reached end of file or stream, no data left to decompress */
+			decompressionResult = 0;
+			break;
+		}
+		{
+			size_t outsize = sizeof outbuf;
+			lz4err = LZ4F_decompress(dctx, outbuf, &outsize, compbuf, &insize, NULL);
+			if (LZ4F_isError(lz4err)) {
+				pr_err("LZ4 Decompress error: %s\n", LZ4F_getErrorName(lz4err));
+				break;
+			}
+
+			totalread += insize;
+			compbuf += insize;
+			if (mapped_size < (compbuf - mapped_addr)) {
+				pr_err("decompression went wrong, compression buffer is out of bounds\n");
+				break;
+			}
+
+			{
+				// Write decompressed bytes
+				ssize_t bytes_written = 0;
+				ssize_t out_offset = 0;
+				do {
+					bytes_written = write(decomp_fd, outbuf + out_offset, outsize);
+					if (0 > bytes_written) {
+						if (EINTR == errno) {
+							continue;
+						}
+						pr_perror ("write error to decompressed file");
+						break;
+					}
+					out_offset += bytes_written;
+					if (outsize < (size_t)bytes_written) {
+						pr_err("written more bytes in total than needed, expected=%lu, written=%lu\n",
+						       (unsigned long)outsize, (unsigned long)bytes_written);
+						break;
+					}
+					outsize -= bytes_written;
+					totalwrite += bytes_written;
+				} while (0 < outsize);
+			}
+			if (0 < outsize) {
+				pr_err("writing is not completed, outsize=%lu\n", (unsigned long)outsize);
+				break;
+			}
+		}
+	}
+	if (0 > close(decomp_fd)) {
+		pr_perror("failed closing decompressed file");
+	}
+	pr_debug("decompression completed, read %lu, wrote %lu\n", (unsigned long)totalread, (unsigned long)totalwrite);
+	return decompressionResult;
+}
+
+static inline int decompress_image2(const char *mapped_addr, size_t mapped_size) {
+	LZ4F_errorCode_t lz4err;
+	LZ4F_decompressionContext_t dctx;
+	int res;
+
+	lz4err = LZ4F_createDecompressionContext(&dctx, LZ4F_VERSION);
+	if (LZ4F_isError(lz4err)) {
+		pr_err("Can't create LZ4 decompression context, %s\n", LZ4F_getErrorName(lz4err));
+		return -1;
+	}
+
+	res = decompress_image3(mapped_addr, mapped_size, dctx);
+
+	lz4err = LZ4F_freeDecompressionContext(dctx);
+	if (LZ4F_isError(lz4err)) {
+		pr_err("LZ4 free context error: %s\n", LZ4F_getErrorName(lz4err));
+	}
```

Comment:
```suggestion
		decompressionResult = -EINVAL;
	}
```

Reply:
Not sure about it. Failing to free decompression context, as well as mapped memory below, doesn't mean decompression fails. A decompressed image can still be open and read when restoring.